Sources/SwiftFormatPrettyPrint/PrettyPrint.swift

```diff
@@ -121,132 +172,92 @@ public class PrettyPrinter {
         forceBreakStack.append(false)
       }
 
-      // The preceding break's offset is added to the indentation of the group. The indent is
-      // incremented from the outer group's indent.
-      let indentValue = indentStack.last ?? 0
-      indentStack.append(indentValue + offset + lastBreakOffset)
-      relativeIndentStack.append(offset + lastBreakOffset)
-      lastBreakOffset = 0
-
     case .close:
-      if isDebugMode {
-        writeCloseGroupDebugMarker()
-      }
       forceBreakStack.removeLast()
-      indentStack.removeLast()
-      let indentValue = relativeIndentStack.popLast() ?? 0
-      // The offset of the last break needs to be adjusted according to its parent group. This is so
-      // the next open token's indent is initialized with the correct value.
-      lastBreakOffset += indentValue
 
     // Create a line break if needed. Calculate the indentation required and adjust spaceRemaining
     // accordingly.
-    case .break(let size, let offset):
-      if isDebugMode {
-        if let forcebreak = forceBreakStack.last, forcebreak {
-          writeBreakDebugMarker(style: .consistent)
-        } else {
-          writeBreakDebugMarker(style: .inconsistent)
+    case .break(let kind, let size):
+      lastBreakKind = kind
+      var mustBreak = forceBreakStack.last ?? false
+      var isContinuation = false
+
+      switch kind {
+      case .open:
+        currentLineIsContinuation = false
+        openDelimiterBreakStack.append(lineNumber)
+        indentationStack.append(configuration.indentation)
+      case .close:
+        guard let matchingOpenLineNumber = openDelimiterBreakStack.popLast() else {
+          fatalError("Unmatched closing break")
         }
+        mustBreak = lineNumber != matchingOpenLineNumber
+        indentationStack.removeLast()
+      case .continue:
+        isContinuation = true
+      case .same:
+        currentLineIsContinuation = false
+      case .reset:
+        mustBreak = currentLineIsContinuation
       }
 
-      // Check if we are forcing breaks within our current group.
-      let forcebreak = forceBreakStack.last ?? false
-
-      if (length > spaceRemaining || forcebreak) && !lastBreakConsecutive {
-        // Check the indentation of the enclosing group.
-        let indentValue = indentStack.last ?? 0
+      if length > spaceRemaining || mustBreak {
+        currentLineIsContinuation = isContinuation
 
-        spaceRemaining = maxLineLength - indentValue - offset
-        write("\n")
+        if !lastBreakConsecutive {
+          writeNewlines(1, discretionary: false)
 
-        lastBreak = true
-        lastBreakConsecutive = true
-        lastBreakOffset = offset
-        lastBreakValue = indentValue + offset
+          lastBreak = true
+          lastBreakConsecutive = true
+        }
       } else {
         if !lastBreakConsecutive {
-          writeSpaces(size)
-          spaceRemaining -= size
-          lastBreakValue = 0
+          enqueueSpaces(size)
```

Comment:
Since breaks include the length of the content following it, aren't we guaranteed that printing the whitespace of a break can never result in trailing whitespace?

Reply:
If a break is immediately followed by a discretionary newline, the length of the break is computed to be equal to its size (from inspecting the token stream output). So if I have a size 1 break, its length is 1, and if there is more than one character left on the line, it would print that space before the following newline.

AFAIK we didn't have that concern with the old model because we used max-line-length _breaks_ instead of newlines to represent discretionary newlines, so the break lengths summed in such a way that this didn't happen.

Reply:
Just to follow up since I thought about it some more: it seems that the reason a break immediately preceding a newline has its own size as the length is because newlines during the scan phase don't go back and fill in length of previous breaks. I did try implementing this earlier on (with the hope of folding newlines completely into breaks), but it caused some other undesirable behavior (IIRC, breaks were firing when they shouldn't have).

That's something we might want to investigate more in the future, but I think for now, moving spaces to a purely enqueuing approach isn't actively _harmful_, since it does fix the issue as breaks/newlines are working today.